adapter/memQuota/memQuota.go

```diff
@@ -0,0 +1,305 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package memQuota provides a simple in-memory quota implementation. It's
+// trivial to set up, but it has various limitations:
+//
+// - Obviously, the data set must be able to fit in memory.
+//
+// - When the mixer crashes/restarts, all quota values are erased.
+//   This means this isn't good for allocation quotas although
+//   it works well enough for rate limits quotas.
+//
+// - Since the data is all memory-resident and there isn't any cross-node
+//   synchronization, this adapter can't be used in an Istio mixer where
+//   a single service can be handled by different mixer instances.
+package memQuota
+
+import (
+	"fmt"
+	"sort"
+	"strconv"
+	"sync"
+	"time"
+
+	"istio.io/mixer/adapter/memQuota/config"
+	"istio.io/mixer/pkg/adapter"
+)
+
+type builder struct{ adapter.DefaultBuilder }
+
+type memQuota struct {
+	sync.Mutex
+
+	// the definitions we know about, immutable
+	definitions map[string]*adapter.QuotaDefinition
+
+	// the counters we track for allocation quotas, protected by lock
+	cells map[string]int64
+
+	// the rolling windows we track for rate quotas, protected by lock
+	windows map[string]*rollingWindow
+
+	// two ping-ponging maps of active dedup ids
+	recentDedup map[string]int64
+	oldDedup    map[string]int64
+
+	// used for reaping dedup ids
+	ticker *time.Ticker
+
+	// indirection to support fast deterministic tests
+	getTick func() int32
+
+	logger adapter.Logger
+}
+
+// we maintain a pool of these for use by the makeKey function
+type keyWorkspace struct {
+	keys   []string
+	buffer []byte
+}
+
+// pool of reusable keyWorkspace structs
+var keyWorkspacePool = sync.Pool{New: func() interface{} { return &keyWorkspace{} }}
+
+var (
+	name = "istio/memQuota"
```

Comment:
I have 2 issues with combining these 2 functions into 1.
1. ratelimit and quota are different aspect `kinds` from a user's perspective.
so 
```
r.RegisterQuota(newQuotaBuilder())
r.RegisterRateLimit(newRateLimitBuilder())
```

2. I think there is very little common between AllocationQuota and RateLimit.
Allocation() is different (windowed, and non-windowed) and release is different. 

If there *is* some common code it can be made into a base class or helper funcs.